### PR TITLE
Spike-based Superior Olivary Complex (SSOC) integration

### PR DESCRIPTION
**Pull request by dgutierrezATC**
In this branch, all the needed functionalities for unpacking the localization information within the AER package have been implemented. The compatibility with the already implemented functions has been kept, so the user can use pyNAVIS with both NAS output files and NAS-SOC output files.

### CODE REVIEW

File: src/pyNAVIS/plots.py [hunk 1]
```diff
@@ -385,3 +386,323 @@ def difference_between_LR(spikes_file, settings, return_data = False, graph_tile
...
+        """
+        Plots the 3D spikegram (also known as raster plot) of the MSO information contained in a LocalizationFile.
+        This is, a graph where the X axis represents neuron IDs, the Y axis means time, and the Z axis represents the frequency channels of the cochlea, and where every spike is plotted as a dot.
+
+        Parameters:
+                localization_file (LocalizationFile): Localization file to plot.
+                settings (MainSettings): Configuration parameters for the file to plot.
+                localization_settings (LocalizationSettings): Localization configuration parameters for the file to plot.
+                dot_size (float): Size of the dots used in the spikegram plot.
+                graph_tile (string, optional): Text that will appear as title for the graph.
```
Comment (jpdominguez) [inline src/pyNAVIS/plots.py:487]: Change graph_tile to graph_title.

File: src/pyNAVIS/plots.py [hunk 2]
```diff
@@ -385,3 +386,323 @@ def difference_between_LR(spikes_file, settings, return_data = False, graph_tile
...
+                text = htmap_ax.text(j, i, mso_activity[i, j],
+                            ha="center", va="center", color="w", fontsize='xx-small')
+        
+        plt.title(graph_title, fontsize='x-large')
+
+        plt.tight_layout()
+        htmap_fig.show()
+
+    @staticmethod
+    def mso_spikegram(localization_file, settings, localization_settings, dot_size = 0.2, graph_tile = 'MSO spikegram', verbose = False):
```
Comment (jpdominguez) [inline src/pyNAVIS/plots.py:477]: Change graph_tile to graph_title

File: src/pyNAVIS/plots.py [hunk 3]
```diff
@@ -385,3 +386,323 @@ def difference_between_LR(spikes_file, settings, return_data = False, graph_tile
...
+        # REPRESENTATION
+        plt.style.use('seaborn-whitegrid')
+        msospk_fig = plt.figure()
+        ax = msospk_fig.add_subplot(111, projection='3d')
+        msospk_fig.canvas.set_window_title(graph_tile)
+
+        # Plot all the spikes stored in the localization_file
+        ax.scatter(localization_file.mso_neuron_ids, localization_file.mso_timestamps, localization_file.mso_channels, s=dot_size)
+            
+        if verbose == True: print('SPIKEGRAM CALCULATION', time.time() - start_time)
```
Comment (jpdominguez) [inline src/pyNAVIS/plots.py:508]: Change to MSO SPIKEGRAM CALCULATION

File: src/pyNAVIS/plots.py [hunk 4]
```diff
@@ -385,3 +386,323 @@ def difference_between_LR(spikes_file, settings, return_data = False, graph_tile
...
+        # Set the axis' limits for each axis according to the LocalizationSettings parameters
+        ax.set_xlim([0, localization_settings.mso_num_neurons_channel])
+        ax.set_ylim([0, localization_file.mso_timestamps[-1]])
+        ax.set_zlim([0, settings.num_channels])
+
+        plt.tight_layout()
+        msospk_fig.show()
+
+    @staticmethod
+    def mso_localization(localization_file, settings, localization_settings, graph_title = "MSO localization estimation", start_at_zero = True, verbose = False):
```
Comment (jpdominguez) [inline src/pyNAVIS/plots.py:526]: Change function name to mso_localization_plot --> also change it in the examples and the rest of the code!

File: src/pyNAVIS/plots.py [hunk 5]
```diff
@@ -385,3 +386,323 @@ def difference_between_LR(spikes_file, settings, return_data = False, graph_tile
...
+
+            last_time += settings.bin_size
+
+            mso_activity[i, :] = spikes
+
+            index_max_activity = np.argmax(mso_activity[i])
+
+            mso_max_activity[i] = index_max_activity
+
+        if verbose == True: print('ACTIVITY MATRIX CALCULATION', time.time() - start_time)
```
Comment (jpdominguez) [inline src/pyNAVIS/plots.py:583]: Change to MSO_LOCALIZATION PLOT CALCULATION

File: src/pyNAVIS/plots.py [hunk 6]
```diff
@@ -385,3 +386,323 @@ def difference_between_LR(spikes_file, settings, return_data = False, graph_tile
...
+        num_mso_events = len(localization_file.mso_timestamps)
+
+        for i in range (0, num_mso_events):
+            # Move the frequency channel from the relative range to an absolute range starting at zero
+            freq_channel = localization_file.mso_channels[i] - localization_settings.mso_start_channel
+            neuron_id = localization_file.mso_neuron_ids[i]
+            # Accumulate the activity for each neuron for each frequency channel according to the LocalizationFIle
+            mso_activity[freq_channel][neuron_id] = mso_activity[freq_channel][neuron_id] + 1
+        
+        if verbose == True: print('ACTIVITY MATRIX CALCULATION', time.time() - start_time)
```
Comment (jpdominguez) [inline src/pyNAVIS/plots.py:428]: Change to MSO HEATMAP CALCULATION

File: src/pyNAVIS/plots.py [hunk 7]
```diff
@@ -385,3 +386,323 @@ def difference_between_LR(spikes_file, settings, return_data = False, graph_tile
...
+        num_mso_events = len(localization_file.mso_timestamps)
+
+        # Calculate the total number of events for each frequency channel and for each neuron ID
+        for i in range (0, num_mso_events):
+            freq_channel = localization_file.mso_channels[i] - localization_settings.mso_start_channel
+            neuron_id = localization_file.mso_neuron_ids[i]
+            # Accumulate the activity for each neuron for each frequency channel according to the LocalizationFIle
+            mso_activity[freq_channel][neuron_id] = mso_activity[freq_channel][neuron_id] + 1
+
+        if verbose == True: print('HISTOGRAM CALCULATION:', time.time() - start_time)
```
Comment (jpdominguez) [inline src/pyNAVIS/plots.py:664]: Change to MSO HISTOGRAM CALCULATION

**Review by jpdominguez — CHANGES_REQUESTED** [review]
Please fix the comments before merging the pull request in the master branch.

File: src/pyNAVIS/functions.py [hunk 8]
```diff
@@ -86,26 +87,74 @@ def check_SpikesFile(spikes_file, settings):
...
-	def adapt_SpikesFile(spikes_file, settings):
+	def adapt_timestamps(timestamps, settings):
 		"""
-		Subtracts the smallest timestamp of the SpikesFile to all of the timestamps contained in the file (in order to start from 0)
+		Subtracts the smallest timestamp of the timestamps list to all of the timestamps contained in the list (in order to start from 0)
 		It also adapts timestamps based on the tick frequency (ts_tick in the MainSettings).
 		
 		Parameters:
-				spikes_file (SpikesFile): File to adapt.
+				timestamps (int[]): File to adapt.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:146]: Done!

[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:146]: Done!

File: src/pyNAVIS/functions.py [hunk 9]
```diff
@@ -266,16 +314,20 @@ def extract_channels_activities(spikes_file, addresses, reset_addresses = True,
...
-	def PDF_report(spikes_file, settings, output_path, plots = ["Spikegram", "Sonogram", "Histogram", "Average activity", "Difference between L/R"], vector = False, verbose = False):
+	def PDF_report(spikes_file, settings, output_path, plots = ["Spikegram", "Sonogram", "Histogram", "Average activity", "Difference between L/R"], add_localization_report = False, localization_file = None, localization_settings = None, localization_plots = ["MSO spikegram", "MSO heatmap", "MSO histogram", "MSO localization"], vector = False, verbose = False):
 		"""
 		Generates a PDF report with the spikegram, sonogram, histogram, average activity and difference between L/R plots obtained from the input SpikesFile or path containing SpikeFiles.
 		
 		Parameters:
 				spikes_file (SpikesFile or string): File or path to use.
 				settings (MainSettings): Configuration parameters for the input file.
+				output_path (string): Destination path.
+				plots (string list): List to select the plots to be included in the PDF report.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:326]: Done!

File: src/pyNAVIS/functions.py [hunk 10]
```diff
@@ -266,16 +314,20 @@ def extract_channels_activities(spikes_file, addresses, reset_addresses = True,
...
+				add_localization_report (boolead, optional): If True, the localization plots will be included.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:327]: Done!

File: src/pyNAVIS/functions.py [hunk 11]
```diff
@@ -266,16 +314,20 @@ def extract_channels_activities(spikes_file, addresses, reset_addresses = True,
...
+				localization_file (LocalizatioFile, optional): If add_localization_report is set to True, this parameter is mandatory, and it should contain the localization information.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:328]: Done!

File: src/pyNAVIS/functions.py [hunk 12]
```diff
@@ -266,16 +314,20 @@ def extract_channels_activities(spikes_file, addresses, reset_addresses = True,
...
+				localization_settings (LocalizationSettings, optional): If add_localization_report is set to True, this parameter is mandatory, and it should contain the localization settings.
+				localization_plots (string list, optional): If add_localization_report is set to True, this parameter is mandatory, and it should contain the list of localization plots.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:330]: Done!

[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:330]: Done!

File: src/pyNAVIS/loaders.py [hunk 13]
```diff
@@ -40,6 +40,30 @@ def __init__(self, addresses = [], timestamps = []):
         self.addresses = addresses
         self.timestamps = timestamps
 
+class LocalizationFile:
+    """
+    Class that contains all the events ant timestamps from the sound source localization model of a file
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/loaders.py:45]: Done!

File: src/pyNAVIS/functions.py [hunk 14]
```diff
@@ -291,14 +343,22 @@ def PDF_report(spikes_file, settings, output_path, plots = ["Spikegram", "Sonogr
 			spikes_file_extension = os.path.splitext(spikes_file)
 
 			if spikes_file_extension == ".aedat":
-				spikes_file = Loaders.loadAEDAT(spikes_file, settings)
+				if add_localization_report == False:
+					spikes_file = Loaders.loadAEDAT(spikes_file, settings)
+				else:
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:348]: Done!

File: src/pyNAVIS/functions.py [hunk 15]
```diff
@@ -291,14 +343,22 @@ def PDF_report(spikes_file, settings, output_path, plots = ["Spikegram", "Sonogr
...
+					spikes_file, localization_file = Loaders.loadAEDATLocalization(spikes_file, settings, localization_settings)
 			elif spikes_file_extension == ".csv":
-				spikes_file = Loaders.loadCSV(spikes_file, settings)
+				if add_localization_report == False:
+					spikes_file = Loaders.loadCSV(spikes_file, delimiter=',')
+				else:
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:353]: Done!

File: src/pyNAVIS/functions.py [hunk 16]
```diff
@@ -291,14 +343,22 @@ def PDF_report(spikes_file, settings, output_path, plots = ["Spikegram", "Sonogr
...
+					spikes_file, localization_file = Loaders.loadCSVLocalization(spikes_file, delimiter=',')
 			elif spikes_file_extension == ".txt":
-				spikes_file = Loaders.loadZynqGrabberData(spikes_file, settings)
+				spikes_file, localization_file = Loaders.loadZynqGrabberData(spikes_file, settings, localization_settings)
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/functions.py:357]: Done!

File: src/pyNAVIS/loaders.py [hunk 17]
```diff
@@ -108,6 +132,153 @@ def loadAEDAT(path, settings):
...
+                            channels_mso.append(freq_channel)
+                            timestamps_mso.append(ts)
+                        elif xso_type == 1:
+                            # LSO event
+                            neuron_ids_lso.append(neuron_id)
+                            channels_lso.append(freq_channel)
+                            timestamps_lso.append(ts)
+                        else:
+                            # Other case
+                            print("[Loaders.loadAEDATLocalization] > DataError: XSO type not recognized!")
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/loaders.py:262]: Done!

File: src/pyNAVIS/loaders.py [hunk 18]
```diff
@@ -140,22 +311,124 @@ def loadCSV(path, delimiter=','):
...
+                        channels_mso.append(freq_channel)
+                        timestamps_mso.append(timestamp)
+                    elif xso_type == 1:
+                        # LSO event
+                        neuron_ids_lso.append(neuron_id)
+                        channels_lso.append(freq_channel)
+                        timestamps_lso.append(timestamp)
+                    else:
+                        # Other case
+                        print("[Loaders.loadCSVLocalization] > DataError: XSO type not recognized!")
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/loaders.py:388]: Done!

File: src/pyNAVIS/loaders.py [hunk 19]
```diff
@@ -164,19 +437,50 @@ def loadZynqGrabberData(path, settings):
...
+                    neuron_ids_mso.append(int(decoded_events_neuron_ids))
+                    
+                elif decoded_events_xso_types == 1:
+                    # LSO event
+                    timestamps_lso.append(int(decoded_events_timestamps))
+                    channels_lso.append(int(decoded_events_freq_ch_addrs))
+                    neuron_ids_lso.append(int(decoded_events_neuron_ids))
+                else:
+                    # Other case
+                    print("[Loaders.loadZynqGrabberData] > DataError: XSO type not recognized!")
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/loaders.py:468]: Done!

File: src/pyNAVIS/main_settings.py [hunk 20]
```diff
@@ -50,4 +50,32 @@ def __init__(self, num_channels, mono_stereo = 0, address_size = 2, ts_tick = 1,
         self.ts_tick = ts_tick
         self.bin_size = bin_size
         self.on_off_both = on_off_both
-        self.reset_timestamp = reset_timestamp
\ No newline at end of file
+        self.reset_timestamp = reset_timestamp
+
+class LocalizationSettings:
+    """
+    Class that collects the configuration settings of pyNAVS when working with a NAS model that integrates the sound source localization model
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/main_settings.py:57]: Done!

File: src/pyNAVIS/main_settings.py [hunk 21]
```diff
@@ -50,4 +50,32 @@ def __init__(self, num_channels, mono_stereo = 0, address_size = 2, ts_tick = 1,
...
+    Attributes:
+        mso_start_channel (int): First NAS frequency channel that is connected to the MSO module.
+        mso_end_channel (int): Last NAS frequency channel that is connected to the MSO module.
+        mso_num_neurons_channel (int): Number of coincidence detector neurons used for each frequency channel in the MSO module.
+        lso_start_channel (int): First NAS frequency channel that is connected to the LSO module.
+        lso_end_channel (int): Last NAS frequency channel that is connected to the LSO module.
+        lso_num_neurons_channel (int): Number of neurons used for each frequency channel in the LSO module.
+
+    Notes:
+            The first frequency channel is the channel number 0.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/main_settings.py:68]: Done!

File: src/pyNAVIS/main_settings.py [hunk 22]
```diff
@@ -50,4 +50,32 @@ def __init__(self, num_channels, mono_stereo = 0, address_size = 2, ts_tick = 1,
...
+
+            The frequency channel range should coincide with the parameter num_channels set in MainSettings.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/main_settings.py:70]: Done!

File: src/pyNAVIS/plots.py [hunk 23]
```diff
@@ -24,7 +24,9 @@
 from bisect import bisect_left, bisect_right
 from matplotlib.colors import LinearSegmentedColormap
 
+import matplotlib
 import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D  # noqa: F401 unused import
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/plots.py:29]: It's being used, but the comment was wrong.

File: src/pyNAVIS/plots.py [hunk 24]
```diff
@@ -385,3 +386,323 @@ def difference_between_LR(spikes_file, settings, return_data = False, graph_tile
...
+        Returns:
+                None.
+
+        Note:
+                None.
+        """
+
+        if verbose == True: start_time = time.time()
+
+        # Get the number of frequency channels set in the configuratio
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/plots.py:413]: Done!

File: src/pyNAVIS/plots.py (hunk 1 above)
Comment (dgutierrezATC) [reply src/pyNAVIS/plots.py:487]: Done!

File: src/pyNAVIS/plots.py (hunk 2 above)
Comment (dgutierrezATC) [reply src/pyNAVIS/plots.py:477]: Done!

File: src/pyNAVIS/plots.py (hunk 3 above)
Comment (dgutierrezATC) [reply src/pyNAVIS/plots.py:508]: Done!

File: src/pyNAVIS/plots.py (hunk 5 above)
Comment (dgutierrezATC) [reply src/pyNAVIS/plots.py:583]: Done!

File: src/pyNAVIS/plots.py (hunk 4 above)
Comment (dgutierrezATC) [reply src/pyNAVIS/plots.py:526]: Done!

File: src/pyNAVIS/plots.py (hunk 6 above)
Comment (dgutierrezATC) [reply src/pyNAVIS/plots.py:428]: Done!

File: src/pyNAVIS/plots.py (hunk 7 above)
Comment (dgutierrezATC) [reply src/pyNAVIS/plots.py:664]: Done!

File: src/pyNAVIS/loaders.py [hunk 25]
```diff
@@ -108,6 +132,153 @@ def loadAEDAT(path, settings):
         spikes_file.timestamps = timestamps
         return spikes_file
 
+    @staticmethod
+    def loadAEDATLocalization(path, settings, localization_settings):
+        """
+        Loads an AEDAT (.aedat) file.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/loaders.py:138]: Done!

File: src/pyNAVIS/loaders.py [hunk 26]
```diff
@@ -140,22 +311,124 @@ def loadCSV(path, delimiter=','):
         spikes_file.timestamps = timestamps
         return spikes_file
 
+    @staticmethod
+    def loadCSVLocalization(path, delimiter=','):
+        """
+        Loads a Comma-Separated Values (.csv) file.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/loaders.py:317]: Done!

File: src/pyNAVIS/main_settings.py [hunk 27]
```diff
@@ -50,4 +50,32 @@ def __init__(self, num_channels, mono_stereo = 0, address_size = 2, ts_tick = 1,
...
+
+            The values set here should be the same that the one used for generating the FPGA configuration file.
```
[the remark this answers is not in the document]
Comment (dgutierrezATC) [reply src/pyNAVIS/main_settings.py:72]: That sentence is to remember the user that both configurations (the one used in Vivado/ISE and here in pyNAVIS should match)

**Pull request merged**